algo: implement sine/cosine based on Intel SVML for single/double precision

seiko2plus opened the pull request:
- Add sin/cos functions with ≤1 ULP and ≤4 ULP accuracy variants
- Support large argument ranges with proper range reduction
- Standalone implementation doesn't require libm
- Written using Google Highway
- implemented for both float32 and float64 data types

seiko2plus committed:
algo: implement sine/cosine using Intel SVML for single/double precision

- Add sin/cos functions with ≤1 ULP and ≤4 ULP accuracy variants
- Support large argument ranges with proper range reduction
- Standalone implementation doesn't require libm
- Written using Google Highway for SIMD optimization
- Optimize for both float32 and float64 data types

seiko2plus committed:
BUG: Fix Precise options and doc comment for it

Mousius reviewed (changes requested):
In general I think this is a good starting point. Do you want to land this with `#if 0` and other various things incomplete @seiko2plus? We can be a bit relaxed as we setup the repo.

seiko2plus committed:
Add code generation tool for SIMD routines

- Add Python-based generator tool for creating C++ headers and Python templates
- Include sollya wrapper for mathematical computations
- Add utilities for formatting C/Python arrays and header generation
- Include comprehensive .gitignore for Python/C++ development

seiko2plus committed:
Add spin build system configuration

- Configure spin commands for building, testing, and development
- Add custom generate command for sollya-based file generation
- Set up IPython integration with pre-imported numpy_sr module
- Configure project metadata and build requirements

seiko2plus committed:
Add SIMD intrinsics testing framework

- Implement Python bindings for Highway SIMD intrinsics
- Add type system for scalar/vector data with proper conversions
- Create dynamic dispatch system for multiple SIMD targets
- Include helper functions for ULP distance calculations
- Set up pytest infrastructure for comprehensive testing

seiko2plus committed:
replace sin/cos lut with the new generators

seiko2plus committed:
Add Highway library as git submodule

Highway provides portable SIMD intrinsics required by the testing framework

seiko2plus committed:
Add npsr meson build placeholder

Initialize build configuration for npsr module

seiko2plus committed:
Add comprehensive stress tests for sin/cos functions

- Generate test vectors using Sollya for high-precision reference values
- Test challenging cases including large arguments, pi multiples, and edge cases
- Cover both float32 and float64 precision with ULP accuracy validation
- Include tests for standard and low-accuracy modes

seiko2plus committed:
Remove Sin/Cos bindings from precision template

Clean up empty PreciseBind function in preparation for proper implementation

seiko2plus committed:
Bind Sin/Cos intrinsics to Python interface

Enable trigonometric functions in the Python testing framework

seiko2plus committed:
Fix polynomial evaluation for large arguments in double precision

- Correct polynomial coefficients for double precision cosine
- Fix table indexing and data layout for lookup tables
- Improve accuracy for large argument range reduction

seiko2plus committed:
Fix small argument handling and add cosine support for double precision

- Implement efficient SVML-style linear approximation for sin/cos
- Fix precision issues in Cody-Waite range reduction
- Add proper sign handling for both sine and cosine
- Improve polynomial evaluation with correction terms
- Document mathematical equivalence between traditional and optimized approaches

seiko2plus committed:
Refactor: Extract Precise class to separate header

- Move Precise class and related types to npsr/precise.h
- Add deduction guides for cleaner template instantiation
- Improve modularity by separating precision control from common utilities

seiko2plus committed:
Refactor: Move Precise parameter to first position in function signatures

- Change parameter order to Precise<Args...>, V for consistency
- Update include paths from sincos/ to trig/
- Fix constant types and naming conventions (kException -> kExceptions)

seiko2plus committed:
Install sin/cos test cases in build system

Add meson configuration to include trigonometric test data in installation

seiko2plus committed:
Add test requirements file

Define Python dependencies for building and testing:
- meson/meson-python for build system
- ninja for compilation
- spin for development workflow
- pytest for test execution

seiko2plus commented on npsr/trig/inl.h:
```diff
@@ -0,0 +1,61 @@
+
+namespace npsr::HWY_NAMESPACE::sincos {
+template <bool IS_COS, typename V, typename Prec>
+HWY_API V SinCos(Prec &prec, V x) {
```
Suggest a name? HWY_API just combine HWY_ATTR and HWY_INLINE.

Mousius replied:
`NPSR_INLINE` which expands to `HWY_ATTR HWY_INLINE` ?

I feel like a lot of things will eventually be using this, so we don't want it to break if Highway's API changes.

seiko2plus replied:
Correction: HWY_API does not include `HWY_ATTR`, since all definitions are placed between the BEFORE/AFTER macros, which emit target attributes through `#pragma`.

`NPSR_INLINE` might cause confusion with `HWY_INLINE`. I’ve updated the PR to use `NPSR_INTRIN` instead — feel free to rename it later if you prefer. Please see:
https://github.com/numpy/numpy-simd-routines/blob/e2e1b10a920f4f15560291ca26cb3652687eb5dc/npsr/hwy.h#L5-L13

Also, please check the latest push for the full set of changes.

Mousius commented:
> All stress tests passed localy on x86, just need a CI and nice cleanup and then it will be ready for review ... maybe I should split it up into multiple prs

Yep, this is unreviewable as-is, so incremental PRs would be much appreciated 😸

r-devulap commented:
> maybe I should split it up into multiple prs

Yes please, I am having a hard time reviewing this.

seiko2plus commented:
> Yep, this is unreviewable as-is, so incremental PRs would be much appreciated
> Yes please, I am having a hard time reviewing this.

Working on it.

seiko2plus committed:
remove unit tests (will be included in a separate PR)

seiko2plus committed:
refactor: reimplement Sollya generator using only Sollya scripts

This patch removes the Python dependency by rewriting the generator using pure Sollya scripting.
It also applies general code cleanups and adds support for non-FMA targets.

seiko2plus committed:
remove auto-generated data (will be added in a separate PR)

seiko2plus committed:
Refactor Sollya integration and enhance trigonometric function precision

- Completely rewrote Sollya integration with a new Python CLI tool (tools/sollya/cli.py)
- Replaced the old generate.py script with a more robust implementation
- Added comprehensive Sollya core library (tools/sollya/core.sol) with utility functions
- Created new SIMD-optimized lookup table implementation (npsr/lut-inl.h)
- Significantly enhanced the Precise class for floating-point precision control
- Refactored trigonometric data generation scripts with better documentation
- Updated constants generation with improved Cody-Waite and Payne-Hanek reduction
- Added kpi16-inl.h.sol for high-precision π/16 table generation
- Removed deprecated utils-inl.h and old trigonometric data files
- Improved project configuration in pyproject.toml and .spin/cmds.py
- Enhanced code documentation throughout the codebase

seiko2plus committed:
DEV: add cross-compiled nix devshells for multiple architectures

- Define `arch_pkgs` mapping for native, aarch64, s390x, ppc64le, and armhf
- Allow selection via `NPSR_ARCH` env var (defaults to native)
- Configure per-arch virtualenv under `.direnv/py/<arch>`
- Bootstrap Python venv with `spin==0.13`

seiko2plus committed:
Move Sollya scripts and Python generator to a separate PR

seiko2plus commented:
Now this PR contains the implementation without data helpers (neither the Sollya scripts nor the generated headers). They can be found in PR #4.

Mousius commented on npsr/trig/inl.h:
```diff
@@ -0,0 +1,151 @@
+ * - Input magnitude (standard vs extended precision for large arguments)
+ * - Special case handling (NaN, Inf)
+ *
+ * @tparam IS_COS   true for cosine, false for sine
```
If this is public, can we use an enum? `true`/`false` is a bit tricky to read.

seiko2plus replied:
Done. It is not suppose to be public, but enum is right call to expand the operations to hold `tan` since the same argument reductions should be reused.

seiko2plus committed:
refactor: replace boolean template parameters with enum-based operation dispatch

- Replace IS_COS boolean template parameters with Operation enum class (kSin/kCos)
- Introduce NPSR_INTRIN macro for consistent intrinsic function attributes
- Update trigonometric function implementations to use new enum-based dispatch
- Improve code readability and maintainability by eliminating boolean parameter confusion
- Update function signatures and documentation to reflect new API design
- Replace HWY_ATTR with HWY_INLINE for better consistency in LUT implementation

Mousius reviewed (approved):
As discussed, let's get things moving and land this and refactor further afterwards 😸 

Thanks for your patience @seiko2plus!
